### PR TITLE
Remove null value pair when parse jsonconfiguration to dictionary .

### PR DESCRIPTION
**Pull request by kklldog**
when read a josn string from consul server like this :
```
{
  "Logging": {
    "LogLevel": {
      "Default": "Information",
      "Microsoft": "Warning",
      "Microsoft.Hosting.Lifetime": "Information"
    }
  },
  "AllowedHosts": "*",
  "server": {
  	"ip":"192.168.0.1",
    "port": 5000
  }
}
```
the JsonConfigurationParser class will parse it to a dictionary . but when json element is an object , it will convert to a null value pair .   
image
    
this pr fix this problem .

### CODE REVIEW

**Comment by kklldog** [comment]
why  AppVeyor build failed  ？

**Comment by Choc13** [comment]
Hi thanks for bringing this up. 

My first instinct though is that this is not a bug, but a feature. I have been caught out by this weird behaviour before so I can see why you would think it is incorrect. However, this is how the `IConfiguration` type works which is defined my Microsoft and not by us. In fact if I remember correctly we are even just using their default `JsonConfigurationParser` so this fix would need to be added there rather than in this library anyway. 

I think the reason for this behaviour is because they overlay this `Section` abstraction on top of this. So I believe a null key relates to a section that only has other sections are children. I think their intention is that you use the `IConfigurationSection` interface to check whether or not is has children before attempting to get the value. 

What particular problem have you encountered? Maybe we can fix this another way as I'm reluctant to deviate from the standard `IConfiguration` behaviour here, even if it is a bit weird. 

Also, I've only been able to check your PR on mobile so I will take a proper look later in case I have misunderstood something.

**Comment by kklldog** [comment]
Hi @Choc13 , 

thanks for answer !
```
 var server = configuration.GetSection("server").Get<Server>() ;
```
Actually your JsonConfigurationParser works fine when i use this code to read server info .

BUT ,
Microsoft's implement 
[JsonConfigurationProvider](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Configuration.Json/src/JsonConfigurationProvider.cs)
[JsonConfigurationFileParser](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Configuration.Json/src/JsonConfigurationFileParser.cs) 
They parse jsonfile to dictionary will not include null value pair . 
That's why i think this is incorrect .
You can try to load a local jsonfile use JsonConfigurationProvider when you are not busy .
My English is not good . I hope we can understand each other . 🤣🤣🤣

**Comment by Choc13** [comment]
Ok yes I think I understand. It sounds like you're saying that the Microsoft libraries also remove these nulls. Let me do a bit of digging through their source code and run some tests to confirm. I'd like to be able to summarise the correct behaviour with a failing test and also make sure we have a test to check that this change doesn't affect model binding from the generated `IConfiguration`.

**Comment by Choc13** [comment]
Also, just to be clear are you saying that `var server = configuration.GetSection("server").Get<Server>();` does work for you?

**Comment by kklldog** [comment]
> Also, just to be clear are you saying that `var server = configuration.GetSection("server").Get<Server>();` does work for you?

That is work for me . The model binding have NO problem .

**Comment by Choc13** [comment]
I've done a little investigation and it appears that the current behaviour is correct. Well, it's at least consistent with what Microsoft's own JSON configuration providers do. 

I ran the following (fsx) script.

```fsharp
#r "nuget: Microsoft.Extensions.Configuration.Json"

open System.IO
open System.Text
open Microsoft.Extensions.Configuration

let config =
    """{ "server": {"ip":"192.168.0.1", "port": 5000} }"""

let stream =
    new MemoryStream(Encoding.UTF8.GetBytes(config))

let parsed =
    ConfigurationBuilder()
        .AddJsonStream(stream)
        .Build()

parsed.AsEnumerable()
```

And the result of `parsed.AsEnumerable()` is 

```
seq
    [[server, ] {Key = "server";
                 Value = null;}; [server:port, 5000] {Key = "server:port";
                                                      Value = "5000";};
     [server:ip, 192.168.0.1] {Key = "server:ip";
                               Value = "192.168.0.1";}]
```

which shows me that the `JsonStreamConfigurationProvider` will return a key with a `null` value for each parent node in the JSON structure.

So unless this is causing you a particular problem with model binding or elsewhere then I think we should leave it the way it is so that it's consistent with the way the other configuration libraries behave when parsing JSON. 

Let me know if you're having any problems though.

**Comment by kklldog** [comment]
hi @Choc13 

```
static void Main(string[] args)
        {
            var config = "{ \"server\":{ \"server\":\"192.168.0.1\", \"port\":5000 }}";

            using (var stream = new MemoryStream(Encoding.UTF8.GetBytes(config)))
            {
                var builder = new ConfigurationBuilder()
                .AddJsonStream(stream)
                .Build();

                //Step 1
                //the builder have only one configuration provider that is JsonConfigurationProvider
                //here , jsonConfigProvder have a property Data(IDictionary) ,
                //it is private , so we can not visit it direct , but use visualstudio we can see the Data not include null value pair .
                var jsonConfigProvder = builder.Providers.First();

                //Step 2
                // AsEnumerable method is the KEY , it will convert the dictionary to IEnumerable that include null pair .
                var kvList = builder.AsEnumerable();
                foreach (var pair in kvList)
                {
                    Console.WriteLine($"{pair.Key} = {pair.Value}");
                }

                //Step 3
                //then convert the IEnumerable to dictionary , the dictionary will include null pair 
                var data = kvList.ToDictionary(pair => pair.Key, pair => pair.Value);
                // the new data(Dictionary) is different from the jsonConfigProvder.Data

                Console.ReadLine();
            }
        }
```
image

These test code will show you , why current JsonConfigurationParser will bring null value pair into the dictionary . 
The KEY is IConfiguartionRoot.AsEnumerable() method will convert the dictionary to IEnumerable that include null pair ,
and IEnumerable.ToDictionary method take the null pair into the new dictionary .
Although current ConsulConfigurationProvider is work for model binding , but there are a little different  between Microsoft's JsonStreamConfiguartionProvider.Data and  ConsulConfigurationProvider.Data .

**Comment by Choc13** [comment]
OK yes I see where the issue is now, it's quite subtle. So [`IConfiguration.AsEnumerable()`](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationExtensions.cs#L52) adds key value pairs for every node in the config hierarchy, including all intermediate parent nodes. 

This library is calling `AsEnumerable` in [JsonConfigurationParser](https://github.com/wintoncode/Winton.Extensions.Configuration.Consul/blob/master/src/Winton.Extensions.Configuration.Consul/Parsers/JsonConfigurationParser.cs#L23). So even though we filter `null` values coming from Consul and go through the `JsonStreamConfigurationProvider` without any modifications, the call to `AsEnumerable` alters the dict as compared to the `Data` property that is inside `JsonStreamConfigurationProvider`.

Unfortunately there is no other public API for getting at the raw `Data` when dealing with an `IConfiguration` object. I think we have 4 options.

1. Do as your PR does and filter the nulls out. This is the easiest fix, however we've coupled ourselves to the internals of how the `AsEnumerable` extension method works, which doesn't seem quite right to.
1. Use reflection to access the `Data` property of the `JsonStreamConfigurationProvider`. I think this is probably worse than option 1 as we're not using the public API at all then.
1. Write our own extension method that is like `AsEnumerable()` but only returns the KVPs for the leaf nodes in the config hierarchy. 
1. Create a class that extends `JsonStreamConfigurationProvider` and exposes the protected `Data` property so that we can access it directly. This would look something like this.

```csharp
public sealed class JsonConfigurationParser : IConfigurationParser
{
    private sealed class JsonStreamParser : JsonStreamConfigurationProvider
    {
        private JsonStreamParser(JsonStreamConfigurationSource source)
            : JsonStreamConfigurationProvider(source)
        {
        }

        internal static IDictionary<string, string> Parse(Stream stream)
        {
            var provider = new JsonStreamParser(new JsonStreamConfigurationSource { Stream = stream });
            provider.Load();
            return provider.Data;
        }
    }

    public IDictionary<string, string> Parse(Stream stream)
    {
        return JsonStreamParser(stream).Parse();
    }
}
```

Although the last option is more code overall I think this is a more correct way to solve this problem because it's very clear then that we're just using the exact same config loading behaviour as the `JsonStreamConfigurationProvider` does.

What do you think? 

Also, would you mind adding a test that would fail with current behaviour, i.e. one that checks that the built dictionary only contains the KVPs for the leaf nodes in the JSON config?

Thanks.

**Comment by kklldog** [comment]
Hi @Choc13 ,
Option 4 is absolutely the MOST correct way !!! It will keep the loading behaviour 100% as the same with Microsoft's code . Great Job !!!

**Comment by Choc13** [comment]
Hi @kklldog did you want to make the change I proposed in this PR? I'm no longer an official maintainer/owner of this repo so I'd also have to PR those changes anyway if I did them. Plus it probably makes sense to do them on this PR so that all the context is in the same place, i.e. the discussion.

**Comment by kklldog** [comment]
Hi @Choc13 ,
I push a new commit .

File: src/Winton.Extensions.Configuration.Consul/Parsers/JsonConfigurationParser.cs
Comment (Choc13) [inline src/Winton.Extensions.Configuration.Consul/Parsers/JsonConfigurationParser.cs:37]: Rather than disable and re-enable the warning I think you can just move this method above the `private sealed class JsonStreamParser` and it will work.

**Comment by Choc13** [comment]
@kklldog thanks for pushing that. Looks good. Could you add some tests that would have failed with the previous implementation to avoid this regression in the future?

Also, I'll try and do a quick PR to move the builds to GitHub actions now because I can't quite figure out what's going on with the AppVeyor and Travis builds tbh (some error to do with GitVersion not checking out the code correctly) and it would be nicer if this was all building on GitHub actions anyway now.

**Comment by Choc13** [comment]
@kklldog I've opened #119 to switch the build over to GH actions. Once that is merged then you should be able to rebase on top of that and the builds will hopefully work fine for you then.

File: src/Winton.Extensions.Configuration.Consul/Parsers/JsonConfigurationParser.cs
Comment (kklldog) [reply src/Winton.Extensions.Configuration.Consul/Parsers/JsonConfigurationParser.cs:37]: agree .

**Comment by Choc13** [comment]
@kklldog you might want to rebase your branch on top of master instead of merging master in. GitHub seems to have gotten a bit confused about what the actual diff is now.

**Review by Choc13 — APPROVED** [review]
Nice work 🙌 Looks good 👀

**Review by lukebailey — APPROVED** [review]

**Pull request merged**
